### PR TITLE
FixedPoint class changes

### LINKED ISSUE #800: fixed point native implementations must replace casting to double throughout math library

### PR DESCRIPTION
**Pull request by markos**
Modify FixedPoint class to avoid conversions to/from doubles, arithmetic is only done in integer.
Add several constants like smallest_fraction, largest_fraction, min, max, etc
Add several tests
Inline all methods by default
Multiplication and division are done in the NextType type which is double the width. In the case of 64-bit types it uses the __int128_t provided by gcc/clang.
It will be possible to avoid this, if we restrict the ranges of the values and lose some accuracy in the fractional parts.
Optimise HighestBitSet method using the __builtin_clzll() function.
Extend tests for FixedPoint<32, 32>
Add tests for FixedPoint<16, 16>

TODO: Add tests for FixedPoint<8, 8>
TODO: Add provision for FixedPoint<128, 128>, this will also need provision of 256-bit arithmetic if we need the full range, or limit the range of the values and lose some accuracy so we do it only using 128bit arithmetic.

Closes #800, #680.

### CODE REVIEW

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (baykaner) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: It may be that I've misunderstood, but this is an implicit cast from a T to a uint64_t? why is this safe? don't we get compiler errors complaining about this?

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (baykaner) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: awesome! didn't know this builtin existed!

File: libs/core/tests/fixed_point/fixed_point.cpp
Comment (baykaner) [inline libs/core/tests/fixed_point/fixed_point.cpp]: is this debugging code you forgot to remove?
I'm a little confused as to why you need a try catch in a test?

File: libs/core/tests/fixed_point/fixed_point.cpp
Comment (baykaner) [inline libs/core/tests/fixed_point/fixed_point.cpp]: There are two valid options with regards to TODOs in the codebase:
1. fix it and remove it
2. raise an issue, assign it to the relevant person, and put " // TODO(issue number): brief description " in the code

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (baykaner) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: nice!

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/core/tests/fixed_point/fixed_point.cpp
Comment (markos) [reply libs/core/tests/fixed_point/fixed_point.cpp]: changed to EXPECT_THROW, thanks :+1:

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: I would suggest do make all these static `const` members rather static `constexpr` (this also applies for all template specializations).

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `static constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `static constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `static constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `static constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `static constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: `const` -> `static constexpr`

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: Just comment: Perhaps I'm mistaken, but `inline` should not be necessary for any methods which are templated or members of templated class or both.
I suppose it does not hurt though.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: Shouldn't here be rather 256 than 128?

And on that note, if `size` data-member will be made `constexpr` as suggested in one of my previous comments, then the next size can be (and shall be) derived in general matter as follows in **every** specialization of this class:

```cpp
static constexpr std::size_t size     = 128;
using NextSize                    = TypeFromSize<(size << 1)>;
// OR:
//using NextSize                    = TypeFromSize<size * 2>;
```

However, since we do not provide implementation for the 256 bit version, we should either do 2 of following things:

1) **EITHER** completely omit declaration/typedefing of the `NextSize` type here inside of  `TypeFromSize<128>` in order to make compiler to fail to compile if someone actually wants to use `TypeFromSize<128>::NextSize` type,

2) **OR** keep `NextSize` typedef in `TypeFromSize<128>`, and rather declare **empty** type for 256 template class specialisation:
```
template<>
struct TypeFromSize<256>;
```

**Review by pbukva — CHANGES_REQUESTED** [review]
See my inline comments

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (bipll) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: It's simply methods defined inside class definition itself are implicitly `inline`.

File: libs/core/tests/fixed_point/fixed_point.cpp
Comment (bipll) [inline libs/core/tests/fixed_point/fixed_point.cpp]: `e` is not mutated inside the `catch` block, so there's no point in catching it by mutable reference.

File: libs/math/include/math/standard_functions/exp.hpp
Comment (bipll) [inline libs/math/include/math/standard_functions/exp.hpp]: Function's ret type is `void` (the second argument to `IfIsFixedPoint`) and `ret` is the return argument.

**Review by bipll — DISMISSED** [review]

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (troels-im) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: This seems like a mistake unless `Data()` gives a reference to the underlying data. I.e. Data() returns a temporary so in effect this swap function is an equality.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (troels-im) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: Const ref rather than ref?

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: I opt for 1, whenever we decide to support FixedPoint<128,128> we will have to deal with a 256-bit type. That's what I did temporarily in the code.

File: libs/core/tests/fixed_point/fixed_point.cpp
Comment (markos) [reply libs/core/tests/fixed_point/fixed_point.cpp]: replaced with EXPECT_THROW

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Good catch, will fix it asap.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Fixed locally.

File: libs/core/tests/fixed_point/fixed_point.cpp
Comment (markos) [reply libs/core/tests/fixed_point/fixed_point.cpp]: Added 843 ticket.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Just did.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: And committed.

File: libs/math/include/math/standard_functions/exp.hpp
Comment (markos) [reply libs/math/include/math/standard_functions/exp.hpp]: Fixed.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: @bipll is correct, inline is only needed outside of the class definition. Removed.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Opted for first choice.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: min still poses a problem, as it gives an 'integer overflow' problem.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Converted.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Not quite, the **only** thing what `inline` keyword actually really guarantees is that there are **NO duplicates** of the symbol during linking process (what would happen **if** the symbol **implementation / definition** is located in **include** file).
This (no symbol duplication = implicit inline) is automatically/implicitly guaranteed for all templated elements (class/struct, type alias, method or function) as long as they are **NOT** fully specialised (they can be **partially** specialised) - this rule is simply implication of the ultimate limitation of C++ templates (whole **implementation / definition** (not just declaration) of templated code **MUST** be delivered exclusively by **including**). 

Just to give a hint, **full** specialisation of templated code should be placed rather in .cpp than .hpp. Depending on which element we fully specialising (struct/class vs. function/method/data-member), **declaration** might need to be present in .hpp (for class/struct/aliased-type), or it does **not** need to be (but can be) declared in .hpp (for  function/method/data-member).

Final physical inlining of method/function is **solely & exclusively** upon compiler to decide whether to do it actually, or not.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Fixed.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: Fixed in last commit.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (baykaner) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: private issues should be labelled: TODO(private 843)

**Review by baykaner — COMMENTED** [review]
This is looking pretty good to me, but there's a lot here and it would be good to get a second pair of eyes on it. @troelsfr can you also take a look?

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pierre-wilmot-fetch-ai) [inline libs/core/include/core/fixed_point/fixed_point.hpp:197]: > The value of std::numeric_limits<T>::max_digits10 is the number of base-10 digits that are necessary to uniquely represent all distinct values of the type T

fractional_bits is number of bits (base 2) we use to store the fractional part.

I think we're comparing apple to oranges here.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (troels-im) [inline libs/core/include/core/fixed_point/fixed_point.hpp]: Please use the wrapper function for pop count in the vectorise library `platform.hpp`. If this particular version does not exist, then it needs to be made. Also, preferably, stick with C++ style casting `static_cast< int32_t >( ... )` as it is clearer.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pierre-wilmot-fetch-ai) [reply libs/core/include/core/fixed_point/fixed_point.hpp:197]: \+ Every call to that function seems to be commented out

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: should we move the platform.hpp inside core? Or is it ok to include it from here?

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pierre-wilmot-fetch-ai) [inline libs/core/include/core/fixed_point/fixed_point.hpp:409]: operator++ uses ONE_MASK
operator-- uses CONST_ONE

That seems like a mistake, if it's not, a comment explaining the inconsitency would be welcome :)

**Review by pierre-wilmot-fetch-ai — DISMISSED** [review]
Looks good, unit test coverage looks pretty intense too.
Since most of the ML library test various arithmetic opeartion using FixedPoint too and Jenkins didn't complain, I assume everything runs OK ;)

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp:409]: that is deliberate, CONST_ONE is actually the FixedPoint object for value of 1.0. ONE_MASK is the integer equivalent.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp:197]: this function could not be used due to constexpr problems and cout. I didn't remove it but it's currently commented out. But you are right that it's wrong to compare max_digits10 and base2 fractional bits. For this reason I will be adding a separate "tolerance" quantity -feel free to suggest another name- per type that will be used for the same purpose. But that will happen in a subsequent PR.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (troels-im) [inline libs/core/include/core/fixed_point/fixed_point.hpp:484]: Please put `{` `}` around

Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp:484]: ack.

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp]: I have added the static_cast, but would prefer to do the platform fix in a different PR.

**Review by pierre-wilmot-fetch-ai — DISMISSED** [review]

**Review by baykaner — DISMISSED** [review]

File: libs/core/include/core/fixed_point/fixed_point.hpp
Comment (pbukva) [inline libs/core/include/core/fixed_point/fixed_point.hpp:719]: Don't fix this now, just something to think about:
The ```% 2``` operation can be done simply as ```& 1``` what should be faster. Perhaps it can be made even faster with little bit of additional wiggling.

Comment (markos) [reply libs/core/include/core/fixed_point/fixed_point.hpp:719]: fixed locally.

**Comment by baykaner** [comment]
I think @bipll is not available to approve the changes he requested but it looks to me like they have been addressed so for the sake of expediency I will dismiss his review.

**Review by pbukva — APPROVED** [review]
LGTM (I made one inline comment, nothing important, just to try something later).

**Pull request merged**
